Perform ISO rendering using a background queue

jayzes opened the pull request:
Instead of our current thread worker implementation. This should theoretically give us the ability to scale iso rendering up/down based on demand, and not have it be tied to the limitations of web worker processes.

Uses https://github.com/Automattic/kue since it has a nice job status API.

TODO:

- [x] Implement queue worker side of things (should just be modding `server-render-entrypoint.js`)
- [x] Add queue worker to Procfile
- [x] Figure out queue length monitoring
- [x] Figure out process concurrency and sizing
- [ ] Add/update tests

Deployment (ninja/epic):
- [ ] Add Redis plan to `ninja`
- [ ] Set `REDIS_PROVIDER` env var
- [ ] Set `CLUSTER_WORKERS` env var (depends on dyno size - maybe 2 for a 2x instance?)
- [ ] Deploy!
- [ ] Scale `queue` worker to 1 2x dyno
- [ ] Scale down web worker to 1 2x dyno
- [ ] Smoke test
- [ ] Add Redis plan to `epic`
- [ ] Set `REDIS_PROVIDER` env var
- [ ] Set `CLUSTER_WORKERS` env var (depends on dyno size - maybe 10 for a PM/PL instance?)
- [ ] Deploy!
- [ ] Scale `queue` worker to 1-2 PM/PL dynos
- [ ] Scale down web worker to 1-2 2x dynos
- [ ] Roll config to `stage1`/`stage2`/`rainbow`
- [ ] Add Librato alerts as needed

jayzes commented:
/cc @rynbyjn

jayzes commented on src/prerender.js:
```diff
@@ -0,0 +1,111 @@
+      Honeybadger.notify(err);
+      done(null, { type: 'error' })
+    })
+    renderToString(InitialComponent)
```
I think this line might might be doubling our work unnecessarily as we call the same thing on line 83

mkitt replied:
Yikes.. sort of looks that way. 💰

jayzes replied:
I'm now not 100% as I just found https://github.com/redux-saga/redux-saga/issues/304 and https://github.com/redux-saga/redux-saga/issues/418 - probably a question for @rynbyjn. I don't understand sagas well enough to understand what's going on there.

ellodanger commented:
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="true">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

jayzes commented:
Found this as well: https://gist.github.com/sompylasar/5e7157e451f4b7268def9ae1ce01edd4#file-serverrenderloop-js

jayzes commented:
@rynbyjn that last commit implements https://www.pivotaltracker.com/story/show/134092451

rynbyjn commented on src/server-iso.js:
```diff
@@ -80,71 +79,66 @@ function saveResponseToCache(cacheKey, body) {
+      access_token: token.token.access_token,
+      originalUrl: req.originalUrl,
+      url: req.url,
+      timingHeader: timingHeader
```
This is the CI failure.. should just be `timingHeader,`

rynbyjn commented on src/selectors/authentication.js:
```diff
@@ -8,14 +8,12 @@ export const selectRefreshToken = state => state.authentication.get('refreshToke
   const expDate = selectExpirationDate(state)
-  const isLoggedIn = selectIsLoggedIn(state)
-  return isLoggedIn && accessToken && expDate > new Date()
+  return accessToken && expDate > new Date()
```
Makes sense and shouldn't effect anything else that I can think of 🤘

jayzes commented:
Okay, think this is ready to go. Adjusted the one bug that @rynbyjn found, corrected the refresh token behavior, cleaned up commits, rebased from master, and re-deployed to rainbow for another set of tests.
